### ISSUE TITLE
GCC support

### ISSUE THREAD

**Issue by LGuerrero13** [cursey/safetyhook#3]
Are there any plans to make this work for GCC so that it can be used in vs code? I've tried linking it just like how the tutorial shows but it seems this was made for Visual Studio in mind, and it throws errors for pegtl with cmake. Is it supposed to only work on the Visual Studio IDE or is there something I am missing?

**Comment by cursey** [comment]
I never actually tried using GCC to compile this project. I've gone ahead and fixed the issues I ran into when doing so in commit 59a4d322e6b41561d6e6a19a355ae00d89745006. Give that a try.
![mintty_eBnpWCvxce](https://user-images.githubusercontent.com/5798459/189503477-6df604f7-c6bc-4b70-b5c2-9e4ab568eed1.png)

**Comment by cursey** [comment]
Also, this project has no dependency on pegtl. Are you sure you made this issue on the right project?

**Comment by LGuerrero13** [comment]
I created a new project with just a main function and nothing else, I then Added this into my CMakeLists.txt file:
`cmake_minimum_required(VERSION 3.0.0)
project(trampolinedll VERSION 0.1.0)

add_library(trampolinedll trampolinedll.cpp)

include(FetchContent)

# Bddisasm
FetchContent_Declare(
    bddisasm
    GIT_REPOSITORY "https://github.com/bitdefender/bddisasm.git"
    GIT_TAG "origin/master"
)
FetchContent_MakeAvailable(bddisasm)

# Safetyhook
FetchContent_Declare(
    safetyhook
    GIT_REPOSITORY "https://github.com/cursey/sdkgenny.git"
    GIT_TAG "origin/master"
)
FetchContent_MakeAvailable(safetyhook)

set(CPACK_PROJECT_NAME ${PROJECT_NAME})
set(CPACK_PROJECT_VERSION ${PROJECT_VERSION})
include(CPack)
`

but I get this output:
`[main] Configuring folder: trampolinedll 
[proc] Executing command: "D:\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" --no-warn-unused-cli -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_C_COMPILER:FILEPATH=C:\mingw64\bin\gcc.exe -DCMAKE_CXX_COMPILER:FILEPATH=C:\mingw64\bin\g++.exe -Sc:/Users/tcclo/Desktop/trampolinedll -Bc:/Users/tcclo/Desktop/trampolinedll/build -G "MinGW Makefiles"
[cmake] Not searching for unused variables given on the command line.
[cmake] -- Extracted version from C:/Users/tcclo/Desktop/trampolinedll/build/_deps/bddisasm-src/inc/version.h: 1.34.18
[cmake] -- Looking for vsnprintf
[cmake] -- Looking for vsnprintf - found
[cmake] -- Looking for memset
[cmake] -- Looking for memset - found
[cmake] CMake Warning at build/_deps/safetyhook-src/CMakeLists.txt:47 (find_package):
[cmake]   By not providing "Findpegtl.cmake" in CMAKE_MODULE_PATH this project has
[cmake]   asked CMake to find a package configuration file provided by "pegtl", but
[cmake]   CMake did not find one.
[cmake] 
[cmake]   Could not find a package configuration file provided by "pegtl" with any of
[cmake]   the following names:
[cmake] 
[cmake]     pegtlConfig.cmake
[cmake]     pegtl-config.cmake
[cmake] 
[cmake]   Add the installation prefix of "pegtl" to CMAKE_PREFIX_PATH or set
[cmake]   "pegtl_DIR" to a directory containing one of the above files.  If "pegtl"
[cmake]   provides a separate development package or SDK, be sure it has been
[cmake]   installed.
[cmake] 
[cmake] 
[cmake] -- Configuring done
[cmake] -- Generating done
[cmake] -- Build files have been written to: C:/Users/tcclo/Desktop/trampolinedll/build
[proc] The command: ctest -N -C Debug failed with error: Error: spawn ctest ENOENT
[ctest] There was an error running ctest to determine available test executables`

**Comment by praydog** [comment]
Looks like we had a typo in the README.md for the example. Sorry about that. Try again.

**Comment by LGuerrero13** [comment]
Now I'm having issues with bddisasm. Currently using mingw64 gcc
![image](https://user-images.githubusercontent.com/58158341/189977600-650f710a-5502-45cf-a1b3-2cc323f3a982.png)

**Comment by angelfor3v3r** [comment]
bddisasm is a tricky one. I've only had luck compiling it with MSVC on windows, it refuses to work with clang-cl (my main compiler).

Could you maybe try installing Visual Studio and use MSVC with VCPKG in CMake (bootstrap it or whatever)? I have no issues compiling bddisasm when I do this.

**Comment by LGuerrero13** [comment]
I got bddisasm working by following the guide on its repo, and building from either the CMake section, or vcpkg section. Even though I fixed bddisasm, I still had other issues with safetyhook where upon calling ` auto builder = factory->acquire(); ` I believe it would just crash (this is on x64) I gave up trying to figure out what was wrong since the implementation looked easy but there was too many issues. Might give it another go though

**Comment by angelfor3v3r** [comment]
That is very odd. I’ve never had it crash ever since I found two bugs a while back that were fixed. If you do try it again and it crashes please post a log or crash dump/debugging info if you don’t mind, thanks!

**Comment by LGuerrero13** [comment]
Here are some issues I'm having currently (using mingw64 gcc). I've linked a video recording of the whole creation of a test program from start to finish [here](https://streamable.com/t3njwo)

1. When including bddisasm into the cmake file, I have to fix the `static __forceinline` to just `__forceinline`. Doesn't seem to play well with static.
2. I cannot include `<SafetyHook.hpp>` directly, I have to follow the relative path from the build folder to access it.
3. When using anything from the SafetyHook file, I get undefined references to things.

This is where I'm currently at when retrying to use this library. BTW I cannot remember how I got it to work somehow last time.

**Comment by angelfor3v3r** [comment]
Thanks! Really helpful information. I had a lot of issues with bddisasm on clang-cl myself because of AES intrinsics. I think most of these issues (mine and yours) stem from bddisasm specifically. Their repo states: “The build process was tested with GCC and Clang on Linux and MSVC on Windows.”. Maybe it would be a good idea to open issues on their site and ask about clang-cl/gcc on windows.

**Comment by angelfor3v3r** [comment]
I’ve opened an issue on bddisasm about this stuff, hopefully it’ll compile on other compilers without hassle soon. I’ll update this issue as I find out more.

**Comment by angelfor3v3r** [comment]
Hi! Could you try again to see if bddisasm compiles without issue now? The nice dev(s) over there added a fix that works for me on clang-cl: https://github.com/bitdefender/bddisasm/commit/102b43dd00e071132d26b1647a04c2655bcf0fce

**Comment by LGuerrero13** [comment]
It seems they've fixed the static issue that was in bddisasm and I also found that including `target_link_libraries(<project name> safetyhook)` into the CMakeList.txt file allows me to include the SafetyHook file regularly as described in the tutorial. There is another issue that appears which is `bddisasm-src/bddisasm/bddisasm.c:4164: undefined reference to __stack_chk_fail` and it seems you have to add `-fstack-protector` as a link flag which removes the error. I do get another problem now with safetyhook this time though which I believe I was having before I gave up last time. Using the example given, whenever I run the example it seems to have a Segmentation fault in the constructor of ThreadFreezer 
![image](https://user-images.githubusercontent.com/58158341/205816716-f8e5696f-7b28-433a-b1ee-ed766dd194ab.png)
![image](https://user-images.githubusercontent.com/58158341/205816931-04e17d4d-7d13-4537-b6f5-2d4732b1508c.png)

**Comment by angelfor3v3r** [comment]
My only guess at this point is something funky is going on with your compiler, etc. I used mingw to compile bddisasm and it’s tools without any warnings. Though, I didn’t test for 32-bit. About the exception: Is there any chance you could try a different debugger or something? I don’t see how an exception could happen *there*. Also could you  post your exact “gcc —version” and how you installed it?

**Comment by LGuerrero13** [comment]
```PS C:\Users\tcclo\Desktop\tester\build> gcc --version
gcc.exe (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
Pretty sure I installed it either through the cygwin installer or from [here](https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/installer/mingw-w64-install.exe/download)

**Comment by LGuerrero13** [comment]
I've finally got it working. It turns out that the crash was specific to the compiler that I was using. I deleted it and decided to download msys2 and install mingw64 g++ from [here](https://www.msys2.org/). Really weird issue, but I guess I fixed it, though one little grime I still have is that I still need to include `target_link_libraries(project safetyhook m)` to be able to include the file without following the relative path from the build folder.
